crates/sandbox/src/lib.rs

```diff
@@ -0,0 +1,360 @@
+    string,
+    time::Duration,
+};
+use tempdir::TempDir;
+use tokio::process::Command;
+use ts_rs::TS;
+
+// -------------------------------------------------------------------------------------------------
+// TYPES
+// -------------------------------------------------------------------------------------------------
+
+pub struct Sandbox {
+    #[allow(dead_code)]
+    scratch: TempDir,
+    input_file: PathBuf,
+    output_dir: PathBuf,
+}
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Unable to create temporary directory: {}", source))]
+    UnableToCreateTempDir { source: io::Error },
+
+    #[snafu(display("Unable to create output directory: {}", source))]
+    UnableToCreateOutputDir { source: io::Error },
+
+    #[snafu(display("Unable to set permissions for output directory: {}", source))]
+    UnableToSetOutputPermissions { source: io::Error },
+
+    #[snafu(display("Unable to create source file: {}", source))]
+    UnableToCreateSourceFile { source: io::Error },
+
+    #[snafu(display("Unable to set permissions for source file: {}", source))]
+    UnableToSetSourcePermissions { source: io::Error },
+
+    #[snafu(display("Output was not valid UTF-8: {}", source))]
+    OutputNotUtf8 { source: string::FromUtf8Error },
+
+    #[snafu(display("Unable to read output file: {}", source))]
+    UnableToReadOutput { source: io::Error },
+
+    #[snafu(display("Unable to start the compiler: {}", source))]
+    UnableToStartCompiler { source: io::Error },
+
+    #[snafu(display("Unable to find the compiler ID"))]
+    MissingCompilerId,
+
+    #[snafu(display("Unable to wait for the compiler: {}", source))]
+    UnableToWaitForCompiler { source: io::Error },
+
+    #[snafu(display("Unable to get output from the compiler: {}", source))]
+    UnableToGetOutputFromCompiler { source: io::Error },
+
+    #[snafu(display("Unable to remove the compiler: {}", source))]
+    UnableToRemoveCompiler { source: io::Error },
+
+    #[snafu(display("Compiler execution took longer than {} ms", timeout.as_millis()))]
+    CompilerExecutionTimedOut {
+        source: tokio::time::error::Elapsed,
+        timeout: Duration,
+    },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[derive(Deserialize, Serialize, TS, Debug, Clone)]
+pub struct CompilationRequest {
+    pub source: String,
+}
+
+#[derive(Deserialize, Serialize, TS, PartialEq, Debug, Clone)]
+#[serde(tag = "type", content = "payload", rename_all = "SCREAMING_SNAKE_CASE")]
+pub enum CompilationResult {
+    Success {
+        wasm: Vec<u8>,
+        stdout: String,
+        stderr: String,
+    },
+    Error {
+        stdout: String,
+        stderr: String,
+    },
+}
+
+// -------------------------------------------------------------------------------------------------
+// CONSTANTS
+// -------------------------------------------------------------------------------------------------
+
+const DOCKER_PROCESS_TIMEOUT_HARD: Duration = Duration::from_secs(12);
```

Comment:
12 seconds might be to small. But this is probably also due to a non up to date docker image due to updating crate.io index and should be resolved when building Smart Contracts with offline flag...